flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/StreamingFileSink.java

```diff
@@ -0,0 +1,397 @@
+		this.bucketCheckInterval = bucketCheckInterval;
+		return this;
+	}
+
+	public StreamingFileSink<IN> setRolloverInterval(long rolloverInterval) {
+		this.rolloverInterval = rolloverInterval;
+		return this;
+	}
+
+	public StreamingFileSink<IN> setInactivityInterval(long inactivityInterval) {
+		this.inactivityInterval = inactivityInterval;
+		return this;
+	}
+
+	@Override
+	public void notifyCheckpointComplete(long checkpointId) throws Exception {
+		final Iterator<Map.Entry<Path, Bucket<IN>>> activeBucketIt =
+				activeBuckets.entrySet().iterator();
+
+		while (activeBucketIt.hasNext()) {
+			Bucket<IN> bucket = activeBucketIt.next().getValue();
+			bucket.commitUpToCheckpoint(checkpointId);
+
+			if (!bucket.isActive()) {
+				// We've dealt with all the pending files and the writer for this bucket is not currently open.
+				// Therefore this bucket is currently inactive and we can remove it from our state.
+				activeBucketIt.remove();
+			}
+		}
+	}
+
+	@Override
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
+		Preconditions.checkNotNull(restoredBucketStates);
+		Preconditions.checkNotNull(fsWriter);
+		Preconditions.checkNotNull(bucketStateSerializer);
+
+		restoredBucketStates.clear();
+		for (Map.Entry<Path, Bucket<IN>> bucketStateEntry : activeBuckets.entrySet()) {
+			final Bucket<IN> bucket = bucketStateEntry.getValue();
+			final Bucket.BucketState bucketState = bucket.snapshot(
+					context.getCheckpointId(),
+					context.getCheckpointTimestamp());
+			restoredBucketStates.add(bucketStateSerializer.serialize(bucketState));
+		}
+
+		restoredMaxCounters.clear();
+		restoredMaxCounters.add(maxPartCounterUsed);
+	}
+
+	@Override
+	public void initializeState(FunctionInitializationContext context) throws Exception {
+		initFileSystemWriter();
+
+		this.activeBuckets = new HashMap<>();
+
+		// Now when restoring, we start fresh. Everything gets committed and the state is empty.
+		// If in the future we want to resume the in-progress files, we should make sure that in
+		// case we receive two states for the same bucket, we merge them appropriately. This includes
+		// keep only one in-progress file and commit the other, and commit the pending ones, as they
+		// were pending for a previous to the last successful checkpoint.
+
+		final OperatorStateStore stateStore = context.getOperatorStateStore();
+
+		restoredBucketStates = stateStore.getListState(bucketStateDesc);
+		restoredMaxCounters = stateStore.getUnionListState(maxPartCounterStateDesc);
+
+		if (context.isRestored()) {
+			final int subtaskIndex = getRuntimeContext().getIndexOfThisSubtask();
+
+			LOG.info("Restoring state for the {} (taskIdx={}).", getClass().getSimpleName(), subtaskIndex);
+
+			for (long partCounter: restoredMaxCounters.get()) {
+				if (partCounter > initMaxPartCounter) {
+					initMaxPartCounter = partCounter;
+				}
+			}
+
+			final int version = bucketStateSerializer.getVersion();
```

Comment:
This needs to use the version of the state serializer that was used to serialize the data, not of the serializer that we happen to have now.

Reply:
I know and I agree, but after a discussion with @StephanEwen, this is a limitation of the current SimpleVersionedSerializer. I may have to somehow checkpoint also the version.

Reply:
It's not a deficiency of the serializer interface, it needs the user to write the version with it. More utilities would be helpful here, agreed.

Reply:
Yes, in this PR that is what is being done. We write the version and read it back.